pkg/genericapiserver/openapi/openapi.go

```diff
@@ -0,0 +1,470 @@
+/*
+Copyright 2015 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openapi
+
+import (
+	"fmt"
+	"net/http"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/emicklei/go-restful"
+	"github.com/emicklei/go-restful/swagger"
+	"github.com/go-openapi/spec"
+)
+
+type Config struct {
+	SwaggerConfig *swagger.Config
+	// List of webservice's path prefixes to ignore
+	IgnorePrefixes []string
+}
+
+type openApi struct {
+	config  *Config
+	swagger *spec.Swagger
+}
+
+func RegisterOpenApiService(config *Config, containers *restful.Container) (err error) {
+	o := openApi{
+		config: config,
+	}
+	err = o.buildSwaggerSpec()
+	if err != nil {
+		return err
+	}
+	containers.ServeMux.HandleFunc("/swagger.json", func(w http.ResponseWriter, r *http.Request) {
+		resp := restful.NewResponse(w)
+		if r.URL.Path != "/swagger.json" {
+			resp.WriteErrorString(http.StatusNotFound, "Path not found!")
+		}
+		resp.WriteAsJson(o.swagger)
+	})
+	return nil
+}
+
+func (o *openApi) buildSwaggerSpec() (err error) {
+	if o.swagger != nil {
+		fmt.Errorf("Swagger spec is already built. Duplicate call to buildSwaggerSpec is not allowed.")
+	}
+	definitions, err := o.buildModels()
+	if err != nil {
+		return err
+	}
+	paths, err := o.buildPaths()
+	if err != nil {
+		return err
+	}
+	o.swagger = &spec.Swagger{
+		SwaggerProps: spec.SwaggerProps{
+			Swagger:     "2.0",
+			Definitions: definitions,
+			Paths:       &paths,
+			Info: &spec.Info{
+				InfoProps: spec.InfoProps{
```

Comment:
Probably not appropriate - this info needs to be passed in.  We're going to have lots of these and not all will strictly be Kubernetes